### PR TITLE
Fixing get_hits_count for es 7.x

### PR DESCRIPTION
**Pull request by JeffAshton**
`get_hits_count` isn't matching in 7.x because of the `doc_type`.  This is an especially bad because it fails to match silently.

### CODE REVIEW

File: elastalert/elastalert.py [hunk 1]
```diff
@@ -456,8 +456,21 @@ def get_hits_count(self, rule, starttime, endtime, index):
             five=rule['five']
         )
 
+        es_client = self.thread_data.current_es
         try:
-            res = self.thread_data.current_es.count(index=index, doc_type=rule['doc_type'], body=query, ignore_unavailable=True)
+            if es_client.is_atleastsixsix():
+                res = self.thread_data.current_es.count(
```
Comment (JeffAshton) [inline elastalert/elastalert.py:462]: Without `doc_type`

File: elastalert/elastalert.py [hunk 2]
```diff
@@ -456,8 +456,21 @@ def get_hits_count(self, rule, starttime, endtime, index):
...
+                    index=index,
+                    body=query,
+                    ignore_unavailable=True
+                )
+            else:
+                res = self.thread_data.current_es.count(
```
Comment (JeffAshton) [inline elastalert/elastalert.py:468]: With `doc_type`:

```
self.thread_data.current_es.count(
                    index=index,
                    doc_type=rule['doc_type'],
                    body={'query': {'match_all': {}}},
                    ignore_unavailable=True
                )
{'count': 0, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}}
```

Comment (JeffAshton) [reply elastalert/elastalert.py:468]: This is silently just not matching any documents in 7.x.  :(

Comment (JeffAshton) [reply elastalert/elastalert.py:468]: Looks to match if you use the implicit `_doc` type, but this makes side by side migrations harder where 6.x clusters had a `doc_type` different from the default.

File: elastalert/elastalert.py [hunk 3]
```diff
@@ -456,8 +456,21 @@ def get_hits_count(self, rule, starttime, endtime, index):
             five=rule['five']
         )
 
+        es_client = self.thread_data.current_es
         try:
-            res = self.thread_data.current_es.count(index=index, doc_type=rule['doc_type'], body=query, ignore_unavailable=True)
+            if es_client.is_atleastsixsix():
+                res = es_client.count(
```
Comment (JeffAshton) [inline elastalert/elastalert.py:462]: Without `doc_type`:

```
es_client.count(
                    index=index,
                    body={'query': {'match_all': {}}},
                    ignore_unavailable=True
                )
{'count': 14074, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}}
```

File: elastalert/elastalert.py [hunk 4]
```diff
@@ -456,8 +456,21 @@ def get_hits_count(self, rule, starttime, endtime, index):
...
+                    index=index,
+                    body=query,
+                    ignore_unavailable=True
+                )
+            else:
```
Comment (JeffAshton) [inline elastalert/elastalert.py:467]: With `doc_type`:

```
es_client.count(
                    index=index,
                    doc_type=rule['doc_type'],
                    body={'query': {'match_all': {}}},
                    ignore_unavailable=True
                )
{'count': 0, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}}
```

Comment (JeffAshton) [reply elastalert/elastalert.py:467]: Looks to match if you use the implicit _doc type, but this makes side by side migrations not possible where 6.x clusters had a doc_type different from the default.

**Review by jertel — COMMENTED** [review]
I noticed no unit test was included. I'm assuming because we're already missing unit coverage for this method?

**Comment by JeffAshton** [comment]
> I noticed no unit test was included. I'm assuming because we're already missing unit coverage for this meth

> I noticed no unit test was included. I'm assuming because we're already missing unit coverage for this method?

I can look at adding some coverage.  Found the issue and wanted to get people aware.   This is one where we really should have a few integration tests with an Elasticsearch docker instance.  Mocking could have easily hidden the same issue.

**Review by jertel — COMMENTED** [review]
Any additional coverage you can provide would be very appreciated. A small improvement to your change would be to define a new helper method called `determineDocType()` which would return `None` when using ES 6.2+. That would serve three purposes:

1. Eliminates maintenance of two similar code statements that call the count() function. 
2. Allows for simple unit testing of the new helper method.
3. Provides a reusable method that we're likely going to need for ES 8.x compatibility.

**Comment by JeffAshton** [comment]
> Any additional coverage you can provide would be very appreciated. A small improvement to your change would be to define a new helper method called `determineDocType()` which would return `None` when using ES 6.2+. That would serve three purposes:
> 
> 1. Eliminates maintenance of two similar code statements that call the count() function.
> 2. Allows for simple unit testing of the new helper method.
> 3. Provides a reusable method that we're likely going to need for ES 8.x compatibility.

I was following the established pattern:

https://github.com/jertel/elastalert2/blob/56c3a2896e453d83671edbfeca95d214ea6fa19d/elastalert/elastalert.py#L1691

There's a few of these in this file.  I think my preference would be to move this branching logic into the Elasticsearch provider.  I wish they hadn't done inheritance, but composition instead, so that all methods used from the external library are veted in one location.

**Pull request merged**